packages/view-types/spatial/src/ScaleBarLayer.js

```diff
@@ -0,0 +1,186 @@
+ * @property {Number} size Physical size of a pixel.
+ * @property {Object} viewState The current viewState for the desired view.  We cannot internally use this.context.viewport because it is one frame behind:
+ * https://github.com/visgl/deck.gl/issues/4504
+ * @property {Array=} boundingBox Boudning box of the view in which this should render.
+ * @property {string=} id Id from the parent layer.
+ * @property {number=} length Value from 0 to 1 representing the portion of the view to be used for the length part of the scale bar.
+ */
+
+/**
+ * @type {{ new(...props: LayerProps[]) }}
+ * @ignore
+ */
+const ScaleBarLayer = class extends CompositeLayer {
+  renderLayers() {
+    const { id, unit, size, position, viewState, length, scaleFactor } = this.props;
+    const boundingBox = makeBoundingBox(viewState);
+    const { zoom } = viewState;
+    const viewLength = boundingBox[2][0] - boundingBox[0][0];
+    const barLength = viewLength * 0.05;
+
+    const adjustment = barLength * scaleFactor;
+
+    // This is a good heuristic for stopping the bar tick marks from getting too small
+    // and/or the text squishing up into the bar.
+    const barHeight = Math.max(
+      2 ** (-zoom + 1.5),
+      (boundingBox[2][1] - boundingBox[0][1]) * 0.007,
+    );
+
+    const numUnits = barLength * size;
+    const [yCoord, xLeftCoord] = getPosition(boundingBox, position, length);
+
+    const lengthBar = new LineLayer({
+      id: `scale-bar-length-${id}`,
+      coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
+      data: [
+        [
+          [xLeftCoord - adjustment, yCoord],
+          [xLeftCoord + barLength, yCoord],
+        ],
+      ],
+      getSourcePosition: d => d[0],
+      getTargetPosition: d => d[1],
+      getWidth: 2,
+      getColor: [220, 220, 220],
+    });
+    const tickBoundsLeft = new LineLayer({
+      id: `scale-bar-height-left-${id}`,
+      coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
+      data: [
+        [
+          [xLeftCoord - adjustment, yCoord - barHeight],
+          [xLeftCoord - adjustment, yCoord + barHeight],
+        ],
+      ],
+
+      getSourcePosition: d => d[0],
+      getTargetPosition: d => d[1],
+      getWidth: 2,
+      getColor: [220, 220, 220],
+    });
+    const tickBoundsRight = new LineLayer({
+      id: `scale-bar-height-right-${id}`,
+      coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
+      data: [
+        [
+          [xLeftCoord + barLength, yCoord - barHeight],
+          [xLeftCoord + barLength, yCoord + barHeight],
+        ],
+      ],
+      getSourcePosition: d => d[0],
+      getTargetPosition: d => d[1],
+      getWidth: 2,
+      getColor: [220, 220, 220],
+    });
+
+    function snapToRound(value) {
+      const magnitude = Math.pow(10, Math.floor(Math.log10(value)));
+      const leadingDigit = Math.round(value / magnitude);
+
+      return leadingDigit * magnitude;
+    }
+
+    const numUnitsRounded = snapToRound(numUnits);
```

Comment:
As far as I understand it, the only thing this rounded number is being used for is the text that is displayed. In order for the bar and the text to match, I think the code to compute length of the bar in pixels needs to depend on the same rounded value that is displayed in the text.

I think the DeckGL viewport project/unproject functions could be used for this https://deck.gl/docs/api-reference/core/viewport#methods

Reply:
I checked out the DeckGL project/unproject documentation that you linked here and I see that we already use `unproject` to calculate the `barLength`. Since I am only manipulating the length of the bar (multiplying it with a scaleFactor number between 0 to 1.5) and this number is later used in the same way as `barLength` before my changes, I don't understand how will I use `project` function and why. Maybe I am missing something?

`As far as I understand it, the only thing this rounded number is being used for is the text that is displayed. ` <-- That is correct -- it is only used for the text. I tried adding this number into the adjustment (I tried variations of this `const adjustment = Math.min(numUnitsRounded * scaleFactor, viewLength * 0.15);`), but I found that at best it doesn't really influence the perception, and usually it looks weird for values of numUnitsRounded < 5 or numUnitsRounded > 500. Maybe in here I am missing out too? I haven't used deck.gl almost at all, so I realise that there can be some concepts here that I am missing.